docs/core/platform/developer_credentials.md

```diff
@@ -0,0 +1,52 @@
+---
+title: "Developer Credentials"
+---
+
+Integrations support [Configuration via OAuth2](https://developers.home-assistant.io/docs/config_entries_config_flow_handler#configuration-via-oauth2) and the preferred approach is to use the Home Assistant Clount Account Linking service. Integrations may also allow users to provide their own developer credentials by adding a `developer_credentials.py` and implementing the right functions.
+
+:::note
+Developer Credentials is under active development and integrations should still prefer using
+`LocalOAuth2Implementation`.
+:::
+
+## Adding support
+
+Integrations support developer credentials with a file in the integration folder called `developer_credentails.py` and implement the following:
+
+```python
+from homeassistant.core import HomeAssistant
+from homeassistant.components.developer_credentials.AuthorizationServer
+from homeassistant.components.developer_credentials.DeveloperCredential
+
+
+async def async_get_authorization_server(
+    self, hass: HomeAssistant
+) -> AuthorizationServer:
+    """Return authorization server."""
+
+
+# Optional and provided only for backwards compatibility if integration used to
+# accept YAML credentials. Return YAML client ID and secret.
+async def async_get_developer_credential(
+    self, hass: HomeAssistant
+) -> DeveloperCredential:
+    """Return a developer credential from configuration.yaml."""
+```
+
+## AuthorizationServer
+
+A `AuthorizationServer` represents the [OAuth2 Authorization server](https://datatracker.ietf.org/doc/html/rfc6749) used for an integration.
+
+| Name          | Type |                                                                                                    | Description |
+| ------------- | ---- | -------------------------------------------------------------------------------------------------- | ----------- |
+| authorize_url | str  | **Required** | The OAuth authorize URL that the user is redirected to during the configuration flow. |
+| token_url     | str  | **Required** | The URL used for obtaining an access token.                                           |
+
+## DeveloperCredential
+
+A `DeveloperCredential` represents the a developer credential provided by the user. This is only provided for backward compatibility by integrations that allowed users to specify developer credentials via `configuration.yaml`.
```

Comment:
Hmm... should integration be able to import and deprecate the YAML configuration?

Reply:
Yes, definitely. This is called out as a task in the core PR, but didn't spell it out here yet:

Options I can see are:
(1) Use the existing platform method:
```
async def async_get_developer_credential(
    self, hass: HomeAssistant
) -> DeveloperCredential:
```
Could also be renamed to have import in the name.

(2) Offer a standalone method for the integration to call that accepts a credentials.

I have a method that in the implementation PR currently private that can be exposed:
```
async def _async_register_auth_implementation(
    hass: HomeAssistant,
    domain: str,
    developer_credential: DeveloperCredential,
) -> None:
```
Or we can rename it (e.g. stick "import" in the name).  Integrations can call this from setup and create a credential from the data in the `config` which would be strictly easier, but more disconnected, from using the platform.
---

Are there gaps in these APIs needed to support import?

I did not mention this here yet because I thought there might be practical implications i have not worked out yet, so was going to work through in an integration first.

Reply:
I was thinking about this more think that import needs a little more from the API, since it also needs to register the authentication domain.  I think I will take the method that returns the credential out of the platform, and make it a static method that the integration calls from setup. 

I next need to take another pass updating the docs to match what is proposed in the PR.

Reply:
I've updated this to reflect a separate import API.  I think it makes sense as a separate method call since new integrations won't need to specify it in the platform.

I was assuming that I didn't need to go to the level of writing out a whole migration guide given the low # of users of the existing local oauth implementation, and that i might be helping with the migration myself anyway.